bug(server): `_revert_to_stored` writes to read-only AppContext properties (#399 phase 1 regression)

## Problem

`_revert_to_stored` in `packages/memtomem/src/memtomem/server/tools/status_config.py:254,256,266` writes to `app.embedder`, `app.search_pipeline`, `app.index_engine`:

```python
def _revert_to_stored(app: AppContext) -> str:
    ...
    new_embedder = create_embedder(config.embedding)
    app.embedder = new_embedder                  # line 254
    app.search_pipeline = SearchPipeline(...)    # line 256
    ...
    app.index_engine = IndexEngine(...)          # line 266
    ...
```

After #399 Phase 1 (`7855042`), these three names on `AppContext` are **read-only properties** that proxy to `self._components.<name>`. Assigning to a `property` without a setter raises `AttributeError: property 'embedder' of 'AppContext' object has no setter` at runtime.

`mypy` flags it:
```
status_config.py:254: error: Property "embedder" defined in "AppContext" is read-only  [misc]
status_config.py:256: error: Property "search_pipeline" defined in "AppContext" is read-only  [misc]
status_config.py:266: error: Property "index_engine" defined in "AppContext" is read-only  [misc]
```

Because `mypy` is advisory in this repo (`CLAUDE.md`) and there is no test that calls `mem_embedding_reset(mode="revert_to_stored")` end-to-end (grep: only `_maybe_offer_embedding_reset` in `tests/test_init_cmd.py`, which hits a different code path), the regression shipped with Phase 1.

## Impact

`mem_embedding_reset(mode="revert_to_stored")` is a recovery path — uncommon, but the whole point is that it stays callable when the DB is in degraded mode (#349). An `AttributeError` in the recovery tool defeats that. The user's fallback is re-starting the server after manually editing config, which is exactly the friction `mem_embedding_reset` exists to remove.

## Fix sketch

`Components` is a mutable `@dataclass` (`server/component_factory.py:30-44`), so the natural fix is to mutate the inner container:

```python
comp = app._components
assert comp is not None  # handler already awaited ensure_initialized
comp.embedder = new_embedder
comp.search_pipeline = SearchPipeline(...)
comp.index_engine = IndexEngine(...)
```

After this, `app.embedder` (the property) reads back the new value — no divergence.

Alternative: add explicit mutator methods on `AppContext` (`replace_embedder` / `replace_search_pipeline` / `replace_index_engine`) so the `_components` access stays behind the class's public surface. Probably overkill for a recovery-only path.

## Test gap to close in the same fix

Add an integration test that calls `mem_embedding_reset(mode="revert_to_stored")` on a fixture in degraded mode and asserts:

1. The tool returns the success message without raising.
2. `app.embedder` / `app.search_pipeline` / `app.index_engine` reflect the stored (not configured) embedder dimension after the call.
3. A follow-up `mem_search` (or any vector-dependent tool) succeeds — demonstrating the three slots were actually swapped, not just `embedder`.

## Context

- Introduced by: #399 Phase 1 (`7855042 feat(server): add AppContext.ensure_initialized plumbing`) — the property conversion of `storage` / `embedder` / `index_engine` / `search_pipeline`. Those properties exist so `AppContext` can defer component creation until first tool call; the read-only shape is deliberate but this one write-path wasn't migrated.
- Surfaced by: #399 Phase 2 handler migration work — `mypy packages/memtomem/src/memtomem/server` output during the PR's lint pass.
- Not a Phase 2 fix because: Phase 2 is handler migration only. Bundling this would mix concerns (`feedback_one_change_per_pr`).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -38,6 +38,17 @@ Format: [Keep a Changelog](https://keepachangelog.com/en/1.1.0/)
 
 ### Fixed
 
+- **`mem_embedding_reset(mode="revert_to_stored")` no longer raises
+  `AttributeError` on the recovery path.** #399 Phase 1 made `embedder`,
+  `search_pipeline`, and `index_engine` read-only `@property`s on
+  `AppContext`, but `_revert_to_stored` kept writing to them directly.
+  Any user hitting the degraded-mode "revert to stored" recovery flow
+  would see `AttributeError: property 'embedder' of 'AppContext' object
+  has no setter` — the exact scenario the tool exists to handle. The
+  writes now mutate `app._components` fields directly (the underlying
+  dataclass is still mutable by design), and a new end-to-end test
+  pins all three runtime slots as swapped, not just `embedder`. (#409)
+
 - **`mm init -y` refuses to write `config.json` when required extras are missing.**
   Previously `-y` accepted `--provider onnx|ollama|openai` and
   `--tokenizer kiwipiepy` without checking whether the corresponding Python

--- a/packages/memtomem/src/memtomem/server/tools/status_config.py
+++ b/packages/memtomem/src/memtomem/server/tools/status_config.py
@@ -250,10 +250,21 @@ def _revert_to_stored(app: AppContext) -> str:
     config.embedding.model = stored["model"]
     config.embedding.dimension = stored["dimension"]
 
+    # ``app.embedder`` / ``app.search_pipeline`` / ``app.index_engine`` are
+    # read-only properties that proxy to ``app._components.<name>`` (#399
+    # Phase 1). Direct assignment would raise ``AttributeError``. The
+    # ``Components`` dataclass is mutable, so we swap fields on the inner
+    # container and the properties pick up the new values automatically.
+    # ``app.storage`` above already dereferenced ``_components``, so the
+    # container is guaranteed non-None by the time we reach here.
+    comp = app._components
+    assert comp is not None, (
+        "_revert_to_stored called before ensure_initialized — "
+        "handler must go through _get_app_initialized"
+    )
     new_embedder = create_embedder(config.embedding)
-    app.embedder = new_embedder
-
-    app.search_pipeline = SearchPipeline(
+    comp.embedder = new_embedder
+    comp.search_pipeline = SearchPipeline(
         storage=storage,
         embedder=new_embedder,
         config=config.search,
@@ -263,7 +274,7 @@ def _revert_to_stored(app: AppContext) -> str:
         context_window_config=config.context_window,
         llm_provider=app.llm_provider,
     )
-    app.index_engine = IndexEngine(
+    comp.index_engine = IndexEngine(
         storage=storage,
         embedder=new_embedder,
         config=config.indexing,

--- a/packages/memtomem/tests/test_server_degraded_mode.py
+++ b/packages/memtomem/tests/test_server_degraded_mode.py
@@ -208,3 +208,45 @@ async def test_mem_embedding_reset_apply_current_repairs_mismatch(degraded_compo
     assert "Embedding mismatch detected" not in message
     assert add_stats is not None
     assert add_stats.indexed_chunks >= 1
+
+
+async def test_mem_embedding_reset_revert_to_stored_swaps_runtime(degraded_components):
+    """Regression for #409: ``revert_to_stored`` mutates ``app._components``
+    fields directly (not the read-only ``AppContext`` properties introduced
+    by #399 Phase 1). Pre-fix this path raised
+    ``AttributeError: property 'embedder' of 'AppContext' object has no setter``
+    the moment it ran, defeating the whole recovery flow.
+
+    The degraded fixture pins stored=none/dim=0, configured=onnx/bge-m3/1024,
+    so reverting downgrades the runtime to a ``NoopEmbedder`` and clears
+    the mismatch. We verify the three runtime slots actually got swapped,
+    not just ``embedder`` — a partial fix that touched only ``embedder``
+    would leave ``search_pipeline`` / ``index_engine`` holding stale
+    references to the configured embedder.
+    """
+    app = _make_app(degraded_components)
+    ctx = _StubCtx(app)
+    pre_embedder = app.embedder
+    pre_search = app.search_pipeline
+    pre_index = app.index_engine
+
+    reset_out = await mem_embedding_reset(mode="revert_to_stored", ctx=ctx)  # type: ignore[arg-type]
+
+    assert "Reverted to stored DB settings" in reset_out
+    assert "none/" in reset_out  # stored provider was "none"
+    assert "0d" in reset_out  # stored dimension was 0
+
+    # All three runtime slots swapped. Identity check is the right assertion:
+    # construction creates a new instance, so the post object is a different
+    # Python object than the pre. Anything narrower (e.g. "dimension == 0")
+    # would silently pass if only ``embedder`` was touched and the pipelines
+    # kept pointing at the old one.
+    assert app.embedder is not pre_embedder
+    assert app.search_pipeline is not pre_search
+    assert app.index_engine is not pre_index
+
+    # Stored-side settings are now reflected in config + live storage view.
+    assert app.config.embedding.provider == "none"
+    assert app.config.embedding.dimension == 0
+    assert app.storage.embedding_mismatch is None
+    assert "DEGRADED" not in await mem_stats(ctx=ctx)  # type: ignore[arg-type]

